source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -4460,42 +4464,51 @@ export class Source {
 }
 
 export class Media extends CardElement {
-    poster: string;
-    altText: string;
-    sources: Array<Source> = [];
     private _mediaType: string;
     private _mediaElement: HTMLMediaElement;
+	private _playButton: HTMLAnchorElement;
+	private _pauseButton: HTMLAnchorElement;
+	private _resetButton: HTMLAnchorElement;
+	private _poster: Image;
+
     static readonly acceptedMediaTypes = ["audio", "video"];
 
-    private processSources(): Array<Source> {
-        let typeFilter = (t): boolean => {
-            if (Media.acceptedMediaTypes.indexOf(t)) {
-                this._mediaType = t;
-                typeFilter = (tt) => tt === this._mediaType;
-                return true;
-            }
-            return false;
-        }
-        return this.sources.filter((s): boolean => {
-            if (s.mimeType) {
-                let mimeComponents = s.mimeType.split('/');
-                if (mimeComponents.length == 2) {
-                    return typeFilter(mimeComponents[0]);
-                }
-            }
-            return false;
-        });
+    private processSources(): Array<MediaSource> {
+		let goodSources: Array<MediaSource> = [];
+
+		for (let source of this.sources) {
+			let mimeComponents = source.mimeType.split('/');
+
+			if (mimeComponents.length == 2) {
+				if (!this._mediaType) {
+					let index = Media.acceptedMediaTypes.indexOf(mimeComponents[0]);
+
+					if (index >= 0) {
+						this._mediaType = Media.acceptedMediaTypes[index];
+					}
+				}
+
+				if (mimeComponents[0] == this._mediaType) {
+					goodSources.push(source);
+				}
+			} 
+		}
+
+		return goodSources;
     }
 
-    private makeButton(label: string): HTMLAnchorElement {
+    private static makeButton(label: string, buttonHint: string, handler: ()=>void): HTMLAnchorElement {
         let button = document.createElement("a");
-        button.className = "ac-media-button";
+        button.classList.add("ac-media-button", buttonHint);
+		button.title = buttonHint;
+		button.onclick = e => { handler(); e.preventDefault(); return false; };
         button.text = label;
+		
         return button;
     }
 
     protected internalRender(): HTMLElement {
-        var element = document.createElement("div");
+        let element = <HTMLElement>document.createElement("div");
```

Comment:
Why cast? It's no necessary.

Reply:
It only becomes unnecessary if I declare the type of the variable explicitly; if I don't, then the type of ```element``` becomes HTMLDivElement and further down for the poster representation I'm invoking render (which returns an HTMLElement); since I prefer to have one rather than two return statements, I need to make the type of 'element' be the base class.

But I think I may remove the concern if I do ```let element: HTMLElement = document.createElement("div")``` instead.

Reply:
I don't think you need to do anything. HTMLDivElement extends HTMLElement, so it *is* an HTMLElement and can be handled naturally as such. In other word you can have a method that is supposed to return an HTMLElement return HTMLDivElement.

Reply:
You got it exactly right, only backwards. If I leave it as 
```let element = document.createElement("div");```
(element remains declared as HTMLDivElement) and further down
```element = something-that-is-HTMLElement-but-possibly-not-HTMLDivElement;```
that other one breaks.

I put it there because tsc was complaining, not because I felt like writing more code today 😃

Reply:
Sorry, I don't get it. There is nothing in your code that requires the cast. It would work just the same without the cast. What is the "complaint" VS Code is giving you?

Reply:
You're completely right. I had the previous round in the back of my head. Sorry for that. Owe you a beer next time I go to Redmond.